### PR TITLE
Primary segment transitions to changetracking when WALrep mirror is down

### PR DESCRIPTION
**Pull request by jimmyyih**
When a WAL replication mirror goes down, FTS should update gp_segment_configuration to mark the mirror down and the primary in changetracking (archiving) mode. While waiting for FTS, the primary segment blocks its commits until it receives transition request to changetracking (archiving) mode. Introduce isolation2 test schedule to validate FTS and WAL replication interactions.

This PR also includes changes to isolation2 to use the gp_inject_fault extension since gpfaultinjector will not work with Segment WAL replication due to the change in fault strategy ('f' removed, 'w' added).  A lot of the Python utilities that utilize the gparray.py library may not work actually (e.g. gpstop, gpcheckcat mirror_matching test).  We skip tests that use these for now when --enable-segwalrep configure flag is set.

### CODE REVIEW

File: src/test/isolation2/sql/setup.sql [hunk 1]
```diff
@@ -90,14 +90,14 @@ SELECT coalesce(
 create language plpythonu;
 -- end_ignore
 
-CREATE OR REPLACE FUNCTION wait_for_trigger_fault(fault text, segno int)
+CREATE OR REPLACE FUNCTION wait_for_trigger_fault(dbname text, fault text, segno int)
```
Comment (ashwinstar) [inline src/test/isolation2/sql/setup.sql]: `CREATE EXTENSION IF NOT EXISTS gp_inject_fault;` is not part of this commit as commit message says but is part of other commit. Can we please bring it to this commit and have this commit as separate PR committed first independently. I am also not sure why create extension needs to be part of setup it can be part of individual tests needing the same right ?

Comment (jimmyyih) [inline src/test/isolation2/sql/setup.sql]: Sure, we can move the commit out as a separate PR and have the `CREATE EXTENSION` in the individual tests.

File: src/backend/cdb/cdbfilerepprimary.c [hunk 2]
```diff
@@ -83,14 +83,21 @@ FileRepPrimary_GetMirrorDataLossTrackingSessionNum(int64 *sessionNum)
```
Comment (jimmyyih) [inline src/backend/cdb/cdbfilerepprimary.c]: Due to postmaster's reliance on pmModuleState for fault transitions, a transition request to changetracking will have the dataState set to `DataStateInChangeTracking` which will return `MirrorDataLossTrackingState_MirrorDown`.  If pmModuleState is not used anywhere, we can remove these changes.

File: src/backend/cdb/cdbwalrep.c [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+/*
+ * Shared-memory variable primaryWalRepState should be consistent with the
+ * WalSndState of the WAL sender in WalSndCtl[0].
+ */
+bool
+isPrimaryWalRepStateConsistent(void)
+{
+	return (getPrimaryWalRepState() == getPrimaryWalRepStateFromWalSnd());
```
Comment (jimmyyih) [inline src/backend/cdb/cdbwalrep.c]: Updated it to hold both spinlocks.

**Comment by jimmyyih** [comment]
@asimrp @ashwinstar this PR has been updated fully after carving out some parts into two separate PRs that have been merged into master.

File: GNUmakefile.in [hunk 4]
```diff
@@ -156,8 +156,25 @@ installcheck-world:
 	$(MAKE) -C src/interfaces/gppc installcheck
 	$(MAKE) -C src/test/kerberos installcheck
 	$(MAKE) -C gpMgmt/bin installcheck
+# The mirroring_matching test of gpcheckcat fails on a walrep cluster
```
Comment (xinzweb) [inline GNUmakefile.in]: How about we have separate commit to let the `gpcheckcat` understand the `w`. That will be better than exclude a test in this way. I tried to keep the GNUmakefile.in simple, and also verify the consistency between primary and mirror. Thoughts?

Seems like this patch might help:
```
diff --git a/gpMgmt/bin/gpcheckcat_modules/mirror_matching_check.py b/gpMgmt/bin/gpcheckcat_modules/mirror_matching_check.py
index 62aa8ed287..3b2b13f7ff 100644
--- a/gpMgmt/bin/gpcheckcat_modules/mirror_matching_check.py
+++ b/gpMgmt/bin/gpcheckcat_modules/mirror_matching_check.py
@@ -7,7 +7,8 @@ class MirrorMatchingCheck:
         logger.info('-----------------------------------')
         logger.info('Checking mirroring_matching')

-        is_config_mirror_enabled = get_gparray_from_config().getFaultStrategy() == FAULT_STRATEGY_FILE_REPLICATION
+        fault_strategy = get_gparray_from_config().getFaultStrategy()
+        is_config_mirror_enabled = fault_strategy in (FAULT_STRATEGY_FILE_REPLICATION, FAULT_STRATEGY_WAL_REPLICATION)

         # This query returns the mirroring status of all segments
         mirroring_query = """SELECT gp_segment_id, mirror_existence_state FROM gp_dist_random('gp_persistent_relation_node') GROUP BY 1,2"""
diff --git a/gpMgmt/bin/gppylib/gparray.py b/gpMgmt/bin/gppylib/gparray.py
index 8217a9aa55..3ed1898719 100755
--- a/gpMgmt/bin/gppylib/gparray.py
+++ b/gpMgmt/bin/gppylib/gparray.py
@@ -109,9 +109,11 @@ def getDataModeLabel(mode):

 FAULT_STRATEGY_NONE = 'n'               # mirrorless systems
 FAULT_STRATEGY_FILE_REPLICATION = 'f'   # valid for versions 4.0+
+FAULT_STRATEGY_WAL_REPLICATION = 'w'   # valid for version 5.0+
 FAULT_STRATEGY_LABELS = {
     FAULT_STRATEGY_NONE:               "none",
     FAULT_STRATEGY_FILE_REPLICATION:   "physical mirroring",
+    FAULT_STRATEGY_WAL_REPLICATION:   "WAL replication",
  }
 VALID_FAULT_STRATEGY = FAULT_STRATEGY_LABELS.keys()```

File: GNUmakefile.in [hunk 5]
```diff
@@ -156,8 +156,25 @@ installcheck-world:
...
+				"part_integrity" "part_constraint" "duplicate_persistent";\
+	do \
+		gpcheckcat -R $$test -A;\
+	done;
+else
 	gpcheckcat -A
 	$(MAKE) -C contrib/pg_upgrade check
+endif
+
+installcheck-walrep:
```
Comment (xinzweb) [inline GNUmakefile.in]: Seems like the `installcheck-resgroup` want to run a subset also.

At this moment, the target we added also duplicate of the target added by `resgroup`. 

Why not just using theirs? Or, maybe better to rename theirs to a better name, like `installcheck-isolation2:`?

File: contrib/pg_upgrade/check.c [hunk 6]
```diff
@@ -891,10 +891,18 @@ check_fts_fault_strategy(migratorContext *ctx)
...
+	/* In segment WAL replication, 'f' is removed and replaced with 'w' */
+	snprintf(query, sizeof(query),
+			 "SELECT fault_strategy "
+			 "FROM   pg_catalog.gp_fault_strategy "
+			 "WHERE  fault_strategy NOT IN ('n','w');");
+#else
 	snprintf(query, sizeof(query),
 			 "SELECT fault_strategy "
 			 "FROM   pg_catalog.gp_fault_strategy "
 			 "WHERE  fault_strategy NOT IN ('n','f');");
```
Comment (xinzweb) [inline contrib/pg_upgrade/check.c]: (with @tvesely): why not just do the minimal change for the NOT IN? say:

```
 	snprintf(query, sizeof(query),
  			 "SELECT fault_strategy "
  			 "FROM   pg_catalog.gp_fault_strategy "
#ifdef USE_SEGWALREP
  			 "WHERE  fault_strategy NOT IN ('n','w');");
#else
  			 "WHERE  fault_strategy NOT IN ('n','f');");
#endif
```

File: src/backend/cdb/cdbfilerepprimary.c (hunk 2 above)
Comment (xinzweb) [inline src/backend/cdb/cdbfilerepprimary.c]: I believe we try to get away from the `pmModuleState`, so, maybe leave this file alone?

File: src/backend/cdb/cdbwalrep.c [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+ * Get the PrimaryWalRepState converted from WalSndCtl.
+ */
+PrimaryWalRepState
+getPrimaryWalRepStateFromWalSnd(void)
+{
+	WalSnd walsender = WalSndCtl->walsnds[0];
+	PrimaryWalRepState state = PRIMARYWALREP_UNKNOWN;
+
+	SpinLockAcquire(&walsender.mutex);
+	if (WalSndCtl->walsnds[0].pid == 0) /* no WAL sender active */
```
Comment (xinzweb) [inline src/backend/cdb/cdbwalrep.c]: Maybe need to revise if we have simplified `PrimaryWalRepState`?

File: src/backend/cdb/cdbwalrep.c [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+		state =  PRIMARYWALREP_STREAMING;
+	SpinLockRelease(&walsender.mutex);
+
+	return state;
+}
+
+const char *
+getPrimaryWalRepStateLabel(PrimaryWalRepState state)
+{
+	static const char *labels[] = {"Archiving",
```
Comment (xinzweb) [inline src/backend/cdb/cdbwalrep.c]: Maybe put this one as `static const` of the file, other than hide it inside this function? Also, we should double check the size of the array is matching the size of the enum.

File: src/backend/cdb/cdbwalrep.c [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+#include "replication/walsender_private.h"
+#include "cdb/cdbwalrep.h"
+
+static volatile PrimaryWalRepState *primaryWalRepState = NULL;
+static slock_t primaryWalRepStateLock;
+
+void
+primaryWalRepStateShmemInit(void)
+{
+	primaryWalRepState = (PrimaryWalRepState *) ShmemAlloc(sizeof(PrimaryWalRepState));
```
Comment (xinzweb) [inline src/backend/cdb/cdbwalrep.c]: Shall we Assert the `primaryWalRepState` is NULL, so that we only initialize it once?

File: GNUmakefile.in (hunk 4 above)
Comment (jimmyyih) [inline GNUmakefile.in]: This is only valid for 6.0, not 5.0.  We currently cannot change Greenplum Python libraries to include segment WAL replication stuff until 5.0 is branched off... unless there's a way to ifdef out Python code.

File: src/backend/cdb/cdbwalrep.c (hunk 7 above)
Comment (jimmyyih) [inline src/backend/cdb/cdbwalrep.c]: I would like to revise it now that synced/notsynced states design has been decided on but the team has already agreed not to and instead have some future commit refactor it.

File: src/backend/cdb/cdbwalrep.c [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+ * Block and wait for FTS to update gp_segment_configuration before
+ * moving on.
+ */
+void
+WalWaitForSegmentConfigurationChange(void)
+{
+	while(1)
+	{
+		PrimaryWalRepState state = getPrimaryWalRepState();
+		if (state < PRIMARYWALREP_FAULT)
```
Comment (xinzweb) [inline src/backend/cdb/cdbwalrep.c]: (with @tvesely) If we are going to hold the WalSnd process, shall we use `set_ps_display` to tell administrators that this WalSnd is blocked and waiting for FTS transition?

File: src/backend/cdb/cdbfilerepprimary.c (hunk 2 above)
Comment (jimmyyih) [inline src/backend/cdb/cdbfilerepprimary.c]: Yes, we are trying to not use pmModuleState but disabling its use needs a separate lengthy PR.  Until then, the postmaster and FTS heavily use pmModuleState and we need to unfortunately add our code alongside its flow.

File: src/backend/cdb/cdbwalrep.c [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+	}
+}
+
+/*
+ * Update primaryWalRepState and block until we get FTS transition. We check
+ * for MASTER_CONTENT_ID because the master and standby master are not
+ * monitored by FTS.
+ */
+void
+WalUpdateStandbyState(PrimaryWalRepState state)
```
Comment (xinzweb) [inline src/backend/cdb/cdbwalrep.c]: (with @tvesely) So far, only two callers of this method (from `walsender.c`), and they are all using `PRIMARYWALREP_FAULT`, hence, this method will wait.

If that's the only case, and we always want to block, then, maybe we don't need to pass in the `state`, e.g.

```
void
WaitingFTSUpdateStandbyStateDueToWALRepFailure()
{
	if (GpIdentity.segindex != MASTER_CONTENT_ID)
	{
		setPrimaryWalRepState(PRIMARYWALREP_FAULT);
		WalWaitForSegmentConfigurationChange();
	}
}
```

File: src/backend/cdb/cdbwalrep.c [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+		 getPrimaryWalRepStateLabel(state));
+
+	*primaryWalRepState = state;
+	SpinLockRelease(&primaryWalRepStateLock);
+}
+
+/*
+ * Block and wait for FTS to update gp_segment_configuration before
+ * moving on.
+ */
```
Comment (xinzweb) [inline src/backend/cdb/cdbwalrep.c]: Maybe state in the comment that, this is based on assumption that FTS will update the state once update the `gp_segment_configuration`, and `WalSender` will be blocked by this method if connection is broken. It took me a while to realize how the FTS coordinate with the `WalSender`.

File: src/backend/fts/ftswalrep.c [hunk 13]
```diff
@@ -0,0 +1,668 @@
+/*-------------------------------------------------------------------------
```
Comment (xinzweb) [inline src/backend/fts/ftswalrep.c]: If we change the state, we probably will simplify this file a lot.

File: GNUmakefile.in (hunk 5 above)
Comment (jimmyyih) [inline GNUmakefile.in]: Good catch.  I combined both targets.

File: src/backend/cdb/cdbwalrep.c (hunk 10 above)
Comment (jimmyyih) [inline src/backend/cdb/cdbwalrep.c]: Using `set_ps_display` would be pretty interesting.  Let's probe the team later to see if we should create a story for this.

File: src/backend/cdb/cdbwalrep.c (hunk 11 above)
Comment (jimmyyih) [inline src/backend/cdb/cdbwalrep.c]: We originally had it like that and changed it to be more flexible for future development (at the time, we thought about blocking on WAL stream establishment and other places in respect to segment recovery).  I think we should leave it flexible for now since our designs are constantly changing and we may possibly need to block for other states in future development stories.

File: src/backend/fts/ftswalrep.c (hunk 13 above)
Comment (jimmyyih) [inline src/backend/fts/ftswalrep.c]: Yes, the synced/notsynced states design will probably remove a good chunk of code in this file.  Otherwise, this file is basically ftsfilerep modified to understand WAL replication.

**Comment by jimmyyih** [comment]
Closing this pull request because of updated primary states and FTS transitioning design.
